test/integration/src/test/java/io/pravega/test/integration/ReadWriteTest.java

```diff
@@ -0,0 +1,262 @@
+        //5  readers -> 5 stream segments ( to have max read parallelism)
+        ScalingPolicy scalingPolicy = ScalingPolicy.fixed(5);
+        StreamConfiguration config = StreamConfiguration.builder().scope(scope)
+                .streamName(STREAM_NAME).scalingPolicy(scalingPolicy).build();
+
+        //create a scope
+        Boolean createScopeStatus = controller.createScope(scope).get();
+        log.info("Create scope status {}", createScopeStatus);
+        //create a stream
+        Boolean createStreamStatus = controller.createStream(config).get();
+        log.info("Create stream status {}", createStreamStatus);
+
+        eventsReadFromPravega = new ConcurrentLinkedQueue<>();
+        stopReadFlag = new AtomicBoolean(false);
+        eventData = new AtomicLong(); //data used by each of the writers.
+        eventReadCount = new AtomicLong(); // used by readers to maintain a count of events.
+
+        ClientFactory clientFactory = new ClientFactoryImpl(scope, controller);
+        //start writing events to the stream with 5 writers
+        log.info("creating {} writers", NUM_WRITERS);
+        List<CompletableFuture<Void>> writerList = new ArrayList<>();
+        for (i = 0; i < NUM_WRITERS; i++) {
+            log.info("starting writer{}", i);
+            writerList.add(startNewWriter(eventData, clientFactory));
+        }
+
+        //create a reader group
+        log.info("Creating Reader group : {}", readerGroupName);
+
+        ReaderGroupManager readerGroupManager = new ReaderGroupManagerImpl(scope, controller, clientFactory);
+        readerGroupManager.createReaderGroup(readerGroupName, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(STREAM_NAME));
+
+        log.info(" reader group name {} ", readerGroupManager.getReaderGroup(readerGroupName).getGroupName());
+        log.info(" reader group scope {}", readerGroupManager.getReaderGroup(readerGroupName).getScope());
+        //create 5 readers
+        log.info("creating {} readers", NUM_READERS);
+        List<CompletableFuture<Void>> readerList = new ArrayList<>();
+        String readerName = "reader" + new Random().nextInt(Integer.MAX_VALUE);
+        //start reading events
+        for (i = 0; i < NUM_READERS; i++) {
+            log.info("starting reader{}", i);
+            readerList.add(startReader(readerName + i, clientFactory, readerGroupName,
+                    eventsReadFromPravega, eventData, eventReadCount, stopReadFlag));
+        }
+
+        log.info("online readers {}", readerGroupManager.getReaderGroup(readerGroupName).getOnlineReaders());
+
+        //wait for writers completion
+        for (i = 0; i < writerList.size(); i++) {
+            log.info("get on writer list {} ", writerList.get(i).get());
+            CompletableFuture.allOf(writerList.get(i));
+        }
+
+        // wait for reads = writes
+        while (NUM_EVENTS != eventsReadFromPravega.size()) {
+            Thread.sleep(5);
```

Comment:
Waiting for  #total no. of events written = #events read and then  set stopReadFlag to true(stop reading events).